File: zig/difference-of-squares/difference_of_squares.zig
pub fn squareOfSum(number: usize) usize {
    // computes the sum of i's from 0 to n and squares it
    var sum: usize = 0;
    // works with zig version 0.11.0-dev.2375+771d07268
    // for (1..number + 1) |i| {
    //     sum += i;
    // }

    var i: usize = number;
    while (i > 0) : (i -= 1) {
        sum += i;
    }
    return sum * sum;
}

pub fn sumOfSquares(number: usize) usize {
    // computes the sum of i^2 from 0 to n
    var sum: usize = 0;
    // works with zig version 0.11.0-dev.2375+771d07268
    // for (1..number + 1) |i| {
    //     sum += i * i;
    // }
    var i: usize = number;
    while (i > 0) : (i -= 1) {
        sum += i * i;
    }
    return sum;
}

pub fn differenceOfSquares(number: usize) usize {
    // computes the difference between the square of sum and sum of squares;
    return squareOfSum(number) - sumOfSquares(number);
}

File: julia/difference-of-squares/difference-of-squares.jl
"Square the sum of the first `n` positive integers"
function square_of_sum(n)
    # imperative
    #sum = 0
	#for i in range(1,stop=n)
    #	sum += i
    #end
 	#sum*sum

    # mathy using the available sum over range function
	#sum(1:n)^2

    # Real Math
	#gauss_sum = (n^2 + n)/2 # Gauss sum for 1 + 2 + ... + n 
	# TypeError: in typeassert, expected Integer, got a value of type Float64
	# division operator / uses float. 
	gauss_sum = n*(n + 1) ÷ 2
	gauss_sum^2
end

"Sum the squares of the first `n` positive integers"
function sum_of_squares(n)
#sum = 0
	#for i in 1:n
    #	sum += i^2
    #end
	#sum

    # sum((1:n) .^ 2)
	#sum(x -> x^2, 0:n)


	pyramidal_number = n*(n+1)*(2n+1)÷6 # n² + (n-1)² + ... (n-n)² 
end

"Subtract the sum of squares from square of the sum of the first `n` positive ints"
function difference(n)
    #square_of_sum(n) - sum_of_squares(n)
	#n*(n+1) * n*(n+1) ÷ 4 - n*(n+1) * (2n+1) ÷6 
	x = n*(n+1)
	y = (2n+1)
	
	#x*(x÷4 - y÷6) # doesn't work :-(
	#(x*x÷2 - x*y÷3)÷2  #works, unless x is outside. 
	
	#use 6/6 to simplyfy equation
	x*(3x - 2y)÷12
end
